libyul/optimiser/RedundantStoreEliminator.cpp

```diff
@@ -0,0 +1,436 @@
+
+			if (instruction == Instruction::MSTORE || instruction == Instruction::MLOAD)
+				op.length = YulString(thirtyTwo);
+			else if (instruction == Instruction::MSTORE8)
+				op.length = YulString(one);
+			else if (
+				instruction == Instruction::REVERT ||
+				instruction == Instruction::RETURN ||
+				instruction == Instruction::KECCAK256 ||
+				instruction == Instruction::LOG0 ||
+				instruction == Instruction::LOG1 ||
+				instruction == Instruction::LOG2 ||
+				instruction == Instruction::LOG3 ||
+				instruction == Instruction::LOG4
+			)
+				op.length = identifierNameIfSSA(_functionCall.arguments.at(1));
+			else if (*instruction == Instruction::EXTCODECOPY)
+				op.length = identifierNameIfSSA(_functionCall.arguments.at(3));
+			else if (
+				instruction == Instruction::CALLDATACOPY ||
+				instruction == Instruction::CODECOPY ||
+				instruction == Instruction::RETURNDATACOPY
+			)
+				op.length = identifierNameIfSSA(_functionCall.arguments.at(2));
+			else
+				op.length = {};
+			return {op};
+		}
+		case Instruction::STATICCALL:
+		case Instruction::CALL:
+		case Instruction::CALLCODE:
+		case Instruction::DELEGATECALL:
+		{
+			size_t arguments = _functionCall.arguments.size();
+			return vector<Operation>{
+				Operation{
+					Location::Memory,
+					Effect::Read,
+					identifierNameIfSSA(_functionCall.arguments.at(arguments - 4)),
+					identifierNameIfSSA(_functionCall.arguments.at(arguments - 3))
+				},
+				// Unknown read includes unknown write.
+				Operation{Location::Storage, Effect::Read, {}, {}},
+				Operation{
+					Location::Memory,
+					Effect::Write,
+					identifierNameIfSSA(_functionCall.arguments.at(arguments - 2)),
+					identifierNameIfSSA(_functionCall.arguments.at(arguments - 1))
+				}
+			};
+		}
+		case Instruction::CREATE:
+		case Instruction::CREATE2:
+			return vector<Operation>{
+				Operation{
+					Location::Memory,
+					Effect::Read,
+					identifierNameIfSSA(_functionCall.arguments.at(1)),
+					identifierNameIfSSA(_functionCall.arguments.at(2))
+				},
+				// Unknown read includes unknown write.
+				Operation{Location::Storage, Effect::Read, {}, {}},
+			};
+		case Instruction::MSIZE:
+			// This is just to satisfy the assert below.
+			return vector<Operation>{};
+		default:
+			yulAssert(
+				evmasm::SemanticInformation::storage(*instruction) ==
+				evmasm::SemanticInformation::None,
+				""
+			);
+			yulAssert(
+				evmasm::SemanticInformation::memory(*instruction) ==
+				evmasm::SemanticInformation::None,
+				""
+			);
+			yulAssert(
+				sideEffects.memory == SideEffects::Effect::None &&
+				sideEffects.storage == SideEffects::Effect::None,
+				""
+			);
+
+			break;
+		}
+	}
+
+	vector<Operation> result;
+	if (sideEffects.memory != SideEffects::Effect::None)
+		result.emplace_back(Operation{Location::Memory, Effect::Read, {}, {}});
+	if (sideEffects.storage != SideEffects::Effect::None)
+		result.emplace_back(Operation{Location::Storage, Effect::Read, {}, {}});
```

Comment:
Looks like this block will only run when the `switch` is skipped (technically it also runs after the `default` in the `switch` but conditions in these `if`s are false in that case). I think it would be more readable to negate the `if`, take `switch` out and put this inside instead.

Reply:
Yep, good idea!